astropy/units/tests/test_quantity.py

```diff
@@ -1003,6 +1003,118 @@ def test_to_string(self):
+            (1234567, "m", "_", "1_234_567.0 m", None, None),
+            (
+                137000000,
+                "lyr",
+                ">+30,.2e",
+                "                     +1.37e+08 lyr",
+                None,
+                None,
+            ),
+            (
+                1.23456789e15,
+                "m/s",
+                ".3e",
+                r"$1.235 \times 10^{15} \; \mathrm{\frac{m}{s}}$",
+                "latex",
+                None,
+            ),
+            (
+                123.456,
+                "km/s",
+                ".2f",
+                r"$123.46 \; \mathrm{\frac{km}{s}}$",
+                "latex",
+                None,
+            ),
+            (
+                123.456,
+                "m/s",
+                ".2f",
+                r"$123.46 \; \mathrm{m\,s^{-1}}$",
+                "latex_inline",
+                None,
+            ),
+            (
+                123.456,
+                "m/s",
+                ".3e",
+                r"$1.235 \times 10^{2} \; \mathrm{\frac{m}{s}}$",
+                "latex",
+                None,
+            ),
+            (
+                123.456,
+                "m/s",
+                ".3e",
+                r"$1.235 \times 10^{2} \; \mathrm{m\,s^{-1}}$",
+                "latex_inline",
+                None,
+            ),
+        ],
+    )
+    def test_format_spec(
+        self, input_value, input_unit, format_spec, expected_result, format, precision
+    ):
+        qscalar = u.Quantity(input_value, input_unit)
+        assert (
+            qscalar.to_string(formatter=format_spec, precision=precision, format=format)
+            == expected_result
+        )
+
+    @pytest.mark.parametrize(
+        "input_quantity, formatter, format, expected_result",
+        [
+            # More explicit formatting
+            (1.2345 * u.kg, lambda x: f"{float(x):.2f}", None, r"1.23 kg"),
+            # Numerical Formatting within LaTeX
+            (
+                1.2345 * u.kg,
+                lambda x: f"{float(x):.2f}",
+                "latex",
+                r"$1.23 \; \mathrm{kg}$",
+            ),
+            # More complex formatting inside LaTeX
+            (
+                35 * u.km / u.s,
+                lambda x: f"\\approx {float(x):.1f}",
+                "latex",
+                r"$\approx 35.0 \; \mathrm{\frac{km}{s}}$",
+            ),
+            # Custom explicit complex number formatter
+            (
+                u.Quantity(2.5 - 1.2j),
+                lambda x: f"({x.real:.2f}{x.imag:+.1f}j)",
+                "latex",
+                r"$(2.50-1.2j) \; \mathrm{}$",
+            ),
+        ],
+    )
+    def test_formatter(self, input_quantity, formatter, format, expected_result):
+        result = input_quantity.to_string(formatter=formatter, format=format)
+        assert result == expected_result
+
+    @pytest.mark.parametrize("format_spec", ["b", "o", "x", "c", "s"])
```

Comment:
We'll need a test where `formatter` is a `dict`. I think it is fine for this to be non-exhaustive! Just one example for generic and one for latex.